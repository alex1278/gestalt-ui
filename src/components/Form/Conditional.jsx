import React from 'react';
import { Field } from 'react-final-form';

// eslint-disable-next-line react/prop-types
const Condition = ({ when, is, children }) => (
  <Field name={when} subscription={{ value: true }}>
    {({ input: { value } }) => (value === is ? children : null)}
  </Field>
);

export default Condition;
