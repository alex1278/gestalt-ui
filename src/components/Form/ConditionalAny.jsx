import React from 'react';
import { Field } from 'react-final-form';

// eslint-disable-next-line react/prop-types
const ConditionAny = ({ when, children }) => (
  <Field name={when} subscription={{ value: true }}>
    {({ input: { value } }) => (value ? children : null)}
  </Field>
);

export default ConditionAny;
