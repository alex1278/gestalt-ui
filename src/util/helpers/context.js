/**
 * generateContextEntityState
 * @param {*} params
 */
export function generateContextEntityState(params) {
  const entity = {
    id: params.environmentId || params.workspaceId || null,
    key: null,
  };

  if (params.workspaceId && params.environmentId) {
    entity.key = 'environments';
  }

  if (params.workspaceId && !params.environmentId) {
    entity.key = 'workspaces';
  }

  // if (params.workspaceId && params.environmentId && params.providerId) {
  //   entity.id = params.providerId;
  //   entity.key = 'providers';
  // }

  return entity;
}
