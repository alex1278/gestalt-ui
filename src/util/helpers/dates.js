import moment from 'moment';

export const getNow = () => moment();

export const getAgo = (num = 0, unit = 'day') =>
  moment().subtract(num, unit);

export const isRecentWithin = (timestamp, num, unit) =>
  moment(timestamp).isBetween(getAgo(num, unit), getNow());
