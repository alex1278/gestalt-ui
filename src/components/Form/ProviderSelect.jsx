import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { orderBy } from 'lodash';
import { Field } from 'react-final-form';
import { composeValidators, required } from 'util/forms';
import { getLastFromSplit } from 'util/helpers/strings';
import SelectField from './Controls/SelectField';
import iconMap from '../../Modules/Providers/config/iconMap';

class ProviderSelect extends Component {
  static propTypes = {
    providers: PropTypes.array.isRequired,
    selectedProvider: PropTypes.object.isRequired,
    setSelectedProvider: PropTypes.func.isRequired,
    name: PropTypes.string,
    label: PropTypes.string,
    form: PropTypes.object.isRequired,
    resetFields: PropTypes.array,
    disabled: PropTypes.bool,
  };

  static defaultProps = {
    name: 'properties.provider.id',
    label: 'Provider',
    resetFields: [],
    disabled: false,
  };

  getProvider(id) {
    const { providers } = this.props;

    return providers.find(p => p.id === id);
  }

  handleSelectedProvider = (e) => {
    const { selectedProvider, setSelectedProvider, name, form, resetFields } = this.props;
    const selection = this.getProvider(e.target.value);

    if (selectedProvider.provider.resource_type !== selection.resource_type && resetFields.length) {
      form.batch(() => {
        resetFields.forEach(f => form.change(f.field, f.value || null));
      });
    }

    // calling redux causes some strange RACE behavior here with form.change where it will not set the value
    // therefore, we need to setTimeout the form.change
    setTimeout(() => form.change(name, e.target.value));
    setSelectedProvider(selection);
  }

  generateMenuItems() {
    const { providers } = this.props;
    return orderBy(providers, 'name').map(item => ({
      key: item.id,
      id: item.id,
      name: item.name,
      // secondaryLabel: item.description,
      leftIcon: item.resource_type ? iconMap(getLastFromSplit(item.resource_type)) : null,
    }));
  }

  render() {
    const { name, label, disabled } = this.props;

    return (
      <Field
        id={`select-provider-${label.toLowerCase()}`}
        component={SelectField}
        name={name}
        label={label}
        menuItems={this.generateMenuItems()}
        itemLabel="name"
        itemValue="id"
        onChange={this.handleSelectedProvider}
        disabled={disabled}
        validate={composeValidators(required())}
        required
        async
      />
    );
  }
}

export default ProviderSelect;
