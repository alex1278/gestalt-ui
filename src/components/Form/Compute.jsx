import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Field } from 'react-final-form';
import { Row, Col } from 'react-flexybox';
import { TextField } from 'components/ReduxFormFields';
import { Slider, FontIcon } from 'react-md';
import { composeValidators, fixInputNumber, fixInputDecimal, min, max, mod, required } from 'util/forms';

class ComputeForm extends Component {
  static propTypes = {
    formValues: PropTypes.object.isRequired,
    form: PropTypes.object.isRequired,
    cpuName: PropTypes.string,
    memName: PropTypes.string,
    cpuMin: PropTypes.number,
    cpuMax: PropTypes.number,
    cpuStep: PropTypes.number,
    cpuStepSlider: PropTypes.number,
    memMin: PropTypes.number,
    memMax: PropTypes.number,
    memStep: PropTypes.number,
    memStepSlider: PropTypes.number,
    cpuDiscrete: PropTypes.bool,
    memDiscrete: PropTypes.bool,
  };

  static defaultProps = {
    cpuName: 'properties.cpus',
    memName: 'properties.memory',
    cpuMin: 0.1,
    cpuMax: 4.0,
    cpuStep: 0.1,
    cpuStepSlider: 0.1,
    memMin: 128,
    memMax: 2048,
    memStep: 1,
    memStepSlider: 32,
    cpuDiscrete: true,
    memDiscrete: true,
  };

  handleFocus = (event) => {
    const { target } = event;

    setTimeout(() => target.select());
  }

  onChangeCPU = (value) => {
    const { form, cpuName, cpuMin } = this.props;
    const formattedValue = value || value >= 0
      ? fixInputDecimal(value)
      : cpuMin;

    if (typeof formattedValue === 'number') {
      form.change(cpuName, formattedValue);
    }
  }

  onChangeMemory = (value) => {
    const { form, memName, memMin } = this.props;
    const formattedValue = value || value >= 0
      ? fixInputNumber(value)
      : memMin;

    if (typeof formattedValue === 'number') {
      form.change(memName, formattedValue);
    }
  }

  render() {
    const {
      formValues,
      cpuName,
      memName,
      cpuMin,
      cpuMax,
      cpuStep,
      cpuStepSlider,
      memMin,
      memMax,
      memStep,
      memStepSlider,
      cpuDiscrete,
      memDiscrete,
    } = this.props;

    return (
      <Row gutter={10}>
        <Col flex={10} xs={8}>
          <Slider
            id={`${cpuName}--cpu`}
            label="CPU"
            min={cpuMin}
            max={cpuMax}
            step={cpuStepSlider}
            leftIcon={<FontIcon>memory</FontIcon>}
            value={formValues.properties.cpus}
            onChange={this.onChangeCPU}
            discrete={cpuDiscrete}
            valuePrecision={1}
          />
        </Col>
        <Col flex={2} xs={4} alignSelf="center">
          <Field
            component={TextField}
            name={cpuName}
            step={cpuStep}
            placeholder="CPU"
            type="number"
            required
            onChange={this.onChangeCPU}
            onFocus={this.handleFocus}
            parse={fixInputDecimal}
            format={fixInputDecimal}
            validate={composeValidators(
              min(cpuMin),
              max(cpuMax),
              required('cpu is required'))
            }
          />
        </Col>
        <Col flex={10} xs={8}>
          <Slider
            id={`${memName}--memory`}
            label="Memory (MB)"
            min={memMin}
            max={memMax}
            step={memStepSlider}
            leftIcon={<FontIcon>sd_card</FontIcon>}
            value={formValues.properties.memory}
            onChange={this.onChangeMemory}
            discrete={memDiscrete}
            discreteTicks={memMin}
            tickWidth={6}
          />
        </Col>
        <Col flex={2} xs={4} alignSelf="center">
          <Field
            component={TextField}
            name={memName}
            step={memStep}
            placeholder="Memory"
            type="number"
            required
            onChange={this.onChangeMemory}
            onFocus={this.handleFocus}
            parse={fixInputNumber}
            format={fixInputNumber}
            validate={composeValidators(
              min(memMin),
              max(memMax),
              mod(memStep),
              required('memory is required'))
            }
          />
        </Col>
      </Row>
    );
  }
}

export default ComputeForm;
