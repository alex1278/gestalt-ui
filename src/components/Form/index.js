import Form from './Form';

export { default as Stepper } from './Stepper/Stepper';
export { default as StepActions } from './Stepper/StepActions';
export { default as Conditional } from './Conditional';
export { default as ConditionalAny } from './ConditionalAny';
export { default as ProviderSelect } from './ProviderSelect';
export { default as TextField } from './Controls/TextField';
export { default as PasswordTextField } from './Controls/PasswordTextField';
export { default as SelectField } from './Controls/SelectField';
export { default as Checkbox } from './Controls/Checkbox';
export { default as AceEditor } from './Controls/AceEditor';
export { default as MultiSelect } from './Controls/MultiSelect';
// export { default as Form } from './Form';
export default Form;
