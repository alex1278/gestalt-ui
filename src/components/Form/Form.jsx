import React, { memo } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Link, withRouter } from 'react-router-dom';
import { FlatButton } from 'components/Buttons';
import { FullPageFooter } from 'components/FullPage';
import SaveIcon from '@material-ui/icons/Save';

const FormStyle = styled.form`
  height: 100%;
  width: 100%;
  ${props => !props.noFooterPadding && 'padding-bottom: 56px'};
  ${props => props.disabled && 'pointer-events: none'};
  ${props => props.disabled && 'opacity: 0.4'};
`;

const Form = memo(({ match, children, disabled, disabledCancel, disabledSubmit, submitTitle, showCancel, cancelTo, noFooterPadding, disableFooter, ...rest }) => (
  <FormStyle
    noValidate
    noFooterPadding={noFooterPadding || disableFooter}
    disabled={disabled}
    {...rest}
  >
    {children}
    {!disableFooter &&
      <FullPageFooter
        leftActions={(
          showCancel &&
          <FlatButton
            disabled={disabledCancel}
            component={Link}
            to={cancelTo}
            label="Cancel"
          />
        )}
        rightActions={(
          <FlatButton
            variant="contained"
            type="submit"
            disabled={disabledSubmit}
            color="primary"
            label={submitTitle}
            icon={<SaveIcon fontSize="small" action="inherit" />}
          />
        )}
      />}
  </FormStyle>
));

Form.propTypes = {
  match: PropTypes.object.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
  disabled: PropTypes.bool,
  disabledCancel: PropTypes.bool,
  disabledSubmit: PropTypes.bool,
  submitTitle: PropTypes.string,
  showCancel: PropTypes.bool,
  cancelTo: PropTypes.string,
  noFooterPadding: PropTypes.bool,
  disableFooter: PropTypes.bool,
};

Form.defaultProps = {
  disabled: false,
  disabledCancel: false,
  disabledSubmit: false,
  submitTitle: 'Submit',
  showCancel: false,
  cancelTo: '/',
  noFooterPadding: false,
  disableFooter: false,
};

export default withRouter(Form);
