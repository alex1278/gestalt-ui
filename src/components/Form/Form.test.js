import React from 'react';
import {
  BrowserRouter as Router,
  Route,
} from 'react-router-dom';
import { renderWithTheme } from '../../../test/helpers';
import Form from './Form';

describe('Components::Form::Form', () => {
  it('mounts with default props', () => {
    const { container } = renderWithTheme(
      <Router>
        <Route>
          <Form>
            <div>ho</div>
          </Form>
        </Route>
      </Router>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly when disabled', () => {
    const { container } = renderWithTheme(
      <Router>
        <Route>
          <Form disabled>
            <div>ho</div>
          </Form>
        </Route>
      </Router>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
