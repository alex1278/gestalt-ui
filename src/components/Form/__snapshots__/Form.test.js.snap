// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components::Form::Form mounts with default props 1`] = `
.c4:not(:disabled) {
  color: white !important;
}

.c5 {
  padding-right: 8px;
}

.c1 {
  position: fixed;
  bottom: 0;
  right: 0;
  min-height: 53px;
  width: 100%;
  width: calc(100% - 56px);
  background-color: #fafafa;
  border-top: 1px solid #eee;
  padding: 8px 24px 8px 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 10;
}

.c1 button,
.c1 a {
  margin-left: 3px;
}

.c2 {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c0 {
  height: 100%;
  width: 100%;
  padding-bottom: 56px;
}

<form
  class="c0"
  novalidate=""
>
  <div>
    ho
  </div>
  <footer
    class="c1"
  >
    <div
      class="c2"
    />
    <div
      class="c3"
    >
      <button
        class="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16 c4"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label-2"
        >
          <span
            class="c5"
          >
            <svg
              action="inherit"
              aria-hidden="true"
              class="MuiSvgIcon-root-30 MuiSvgIcon-fontSizeSmall-37"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
              />
            </svg>
          </span>
          Submit
        </span>
        <span
          class="MuiTouchRipple-root-39"
        />
      </button>
    </div>
  </footer>
</form>
`;

exports[`Components::Form::Form renders correctly when disabled 1`] = `
.c4:not(:disabled) {
  color: white !important;
}

.c5 {
  padding-right: 8px;
}

.c1 {
  position: fixed;
  bottom: 0;
  right: 0;
  min-height: 53px;
  width: 100%;
  width: calc(100% - 56px);
  background-color: #fafafa;
  border-top: 1px solid #eee;
  padding: 8px 24px 8px 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 10;
}

.c1 button,
.c1 a {
  margin-left: 3px;
}

.c2 {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c0 {
  height: 100%;
  width: 100%;
  padding-bottom: 56px;
  pointer-events: none;
  opacity: 0.4;
}

<form
  class="c0"
  disabled=""
  novalidate=""
>
  <div>
    ho
  </div>
  <footer
    class="c1"
  >
    <div
      class="c2"
    />
    <div
      class="c3"
    >
      <button
        class="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-contained-12 MuiButton-containedPrimary-13 MuiButton-raised-15 MuiButton-raisedPrimary-16 c4"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label-2"
        >
          <span
            class="c5"
          >
            <svg
              action="inherit"
              aria-hidden="true"
              class="MuiSvgIcon-root-30 MuiSvgIcon-fontSizeSmall-37"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
              />
            </svg>
          </span>
          Submit
        </span>
        <span
          class="MuiTouchRipple-root-39"
        />
      </button>
    </div>
  </footer>
</form>
`;
